spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerTemplate.java

```diff
@@ -145,46 +154,36 @@ public SpannerConverter getSpannerConverter() {
 	}
 
 	@Override
-	public <T> List<T> queryAll(Class<T> entityClass, Sort sort) {
-		return queryAll(entityClass, sort, null);
-	}
-
-	@Override
-	public <T> List<T> queryAll(Class<T> entityClass, Sort sort,
-			SpannerQueryOptions options) {
-		Assert.notNull(sort, "sort must not be null!");
-
-		StringBuilder stringBuilder = new StringBuilder();
+	public <T> List<T> queryAll(Class<T> entityClass, SpannerQueryOptions options) {
 		SpannerPersistentEntity<?> persistentEntity = this.mappingContext
 				.getPersistentEntity(entityClass);
-		stringBuilder.append("SELECT * FROM " + persistentEntity.tableName() + " ");
-		SpannerStatementQueryExecutor.buildOrderBy(persistentEntity, stringBuilder, sort);
-		if (options != null) {
+		String sql = "SELECT * FROM " + persistentEntity.tableName();
+		return query(entityClass, applySortingPagingQueryOptions(options, sql), null,
+				null, options);
+	}
+
+	public String applySortingPagingQueryOptions(SpannerQueryOptions options,
+			String sql) {
+		StringBuilder r = new StringBuilder(applySort(options.getSort(), sql));
 			if (options.hasLimit()) {
-				stringBuilder.append(" LIMIT " + options.getLimit());
+				r.append(" LIMIT " + options.getLimit());
 			}
 			if (options.hasLimit()) {
-				stringBuilder.append(" OFFSET " + options.getOffset());
+				r.append(" OFFSET " + options.getOffset());
 			}
-		}
-		stringBuilder.append(";");
-		return query(entityClass, Statement.of(stringBuilder.toString()), options);
+		return r.toString();
 	}
 
-	@Override
-	public <T> Page<T> queryAll(Class<T> entityClass, Pageable pageable,
-			SpannerQueryOptions options) {
-		Assert.notNull(pageable, "Pageable must not be null!");
-
-		Long count = count(entityClass);
-		List<T> list = queryAll(entityClass, pageable.getSort(), options);
-		return new PageImpl(list, pageable, count);
-	}
-
-	@Override
-	public <T> Page<T> queryAll(Class<T> entityClass, Pageable pageable) {
-		return queryAll(entityClass, pageable, new SpannerQueryOptions()
-				.setOffset(pageable.getOffset()).setLimit(pageable.getPageSize()));
+	private String applySort(Sort sort, String sql) {
+		String s = "SELECT * FROM (" + sql + ")";
+		if (sort == null || sort.isUnsorted()) {
+			return s;
```

Comment:
Why not just return `sql`? The additional subquery just adds an unnecessary performance hit.

Reply:
Are you sure that it does? It seems like something that could be optimized away.

This is needed because the user might already have had a LIMIT and OFFSET at the end of the query and our stated policy in the issue is to apply this on top of given query.

Reply:
To clarify, applySort is used by applySortingPagingQueryOptions, and even if there is no sort the LIMIT and OFFSET bits need to be applied by applySortingPagingQueryOptions after the inner SQL is surrounded by parentheses.

Reply:
It seems like the wrapping into a subquery then maybe should be done before the call to `applySort`.

Reply:
The downside to that is then applySort  cannot be used by the user without the user wrapping the SQL themselves.

Reply:
But this is a private method. How would the user ever call it directly?

Reply:
By "user" i meant any other function that may call this function. It doesn't feel clean to need to mess with the SQL string if you're about to call a function that also touches the contents of this SQL string. I prefer to keep all of the string-touching regarding sorting and pageable within these two functions.